File: androidApp/src/main/java/org/idos/app/security/EthSigner.kt
package org.idos.app.security

import org.idos.app.data.StorageManager
import org.idos.crypto.BouncyCastleKeccak256
import org.idos.crypto.eip712.Eip712Utils
import org.idos.crypto.eip712.TypedData
import org.idos.kwil.types.HexString
import org.kethereum.bip32.toKey
import org.kethereum.bip39.model.MnemonicWords
import org.kethereum.bip39.toSeed
import org.kethereum.crypto.signMessage
import org.kethereum.crypto.signMessageHash
import org.kethereum.crypto.toAddress
import org.kethereum.crypto.toECKeyPair
import org.kethereum.crypto.toHex
import org.kethereum.eip191.signWithEIP191PersonalSign
import org.kethereum.model.PrivateKey

class EthSigner(
    private val keyManager: KeyManager,
    private val storageManager: StorageManager,
) : org.idos.signer.EthSigner(BouncyCastleKeccak256()) {
    override fun getIdentifier(): HexString {
        val address = storageManager.getStoredWallet()?.cleanHex
        requireNotNull(address)
        return address
    }

    override suspend fun sign(msg: ByteArray): ByteArray =
        keyManager.getStoredKey()?.run {
            val signature =
                PrivateKey(this)
                    .toECKeyPair()
                    .signWithEIP191PersonalSign(msg)
                    .toHex()
                    .removePrefix("0x")
                    .hexToByteArray()
            this.fill(0)
            signature
        } ?: byteArrayOf()

    override suspend fun signTypedData(typedData: TypedData): String =
        keyManager.getStoredKey()?.run {
            val keyPair = PrivateKey(this).toECKeyPair()

            // Hash according to EIP-712 using injected utils
            val hash = Eip712Utils.hashTypedData(keccak256, typedData)

            // Sign the raw hash (no personal sign prefix for EIP-712)
            val signature = signMessageHash(hash, keyPair)

            // Clear sensitive data
            this.fill(0)

            // Convert to bytes
            signature.toHex()
        } ?: ""

    companion object {
        /**
         * Standard Ethereum BIP44 derivation path matching iOS
         * m/44'/60'/0'/0/47
         */
        const val DEFAULT_DERIVATION_PATH = "m/44'/60'/0'/0/47"

        fun String.mnemonicToKeypair(derivationPath: String): ByteArray {
            val mnemonic = MnemonicWords(this)
            val seed = mnemonic.toSeed("")
            val key = seed.toKey(derivationPath)
            return key.keyPair.privateKey.key
                .toByteArray()
        }

        fun ByteArray.privateToAddress() =
            PrivateKey(this)
                .toECKeyPair()
                .publicKey
                .toAddress()
    }
}

File: iosApp/iosApp/Security/EthSigner.swift
import Foundation
import WalletCore
import idos_sdk

/// EthSigner provides Ethereum wallet functionality matching Android's EthSigner.kt
/// Handles BIP39 mnemonic derivation, key management, and EIP-191 signing
class EthSigner: idos_sdk.EthSigner {
    private let keyManager: KeyManager
    private let storageManager: StorageManager

    /// Standard Ethereum BIP44 derivation path matching Android
    /// m/44'/60'/0'/0/47
    static let defaultDerivationPath = "m/44'/60'/0'/0/47"

    enum EthSignerError: Error {
        case invalidMnemonic
        case invalidPrivateKey
        case keyDerivationFailed
        case signingFailed
        case noStoredKey
    }

    init(keyManager: KeyManager, storageManager: StorageManager, hasher: Keccak256Hasher) {
        self.keyManager = keyManager
        self.storageManager = storageManager
        super.init(keccak256: hasher)
    }

    // MARK: - EthSigner Override Methods

    /// Get Ethereum address identifier
    /// Matches Android's EthSigner.getIdentifier()
    override func getIdentifier() -> String {
        let address = storageManager.getStoredWallet()
        guard let address = address else {
            fatalError("Wallet address not found in storage")
        }
        return address.removingPrefix("0x")
    }

    /// Sign message with EIP-191 personal sign
    /// Matches Android's EthSigner.sign()
    ///
    /// - Parameter msg: Message bytes to sign as KotlinByteArray
    /// - Returns: 65-byte signature as KotlinByteArray
    override func __sign(msg: KotlinByteArray) async throws -> KotlinByteArray {
        // Convert KotlinByteArray to Data
        let message = msg.toNSData()

        // Get stored private key
        guard let privateKeyData = try? keyManager.getKey() else {
            throw EthSignerError.noStoredKey
        }

        defer {
            // Clear private key from memory
            var mutableKey = privateKeyData
            mutableKey.withUnsafeMutableBytes { ptr in
                if let baseAddress = ptr.baseAddress {
                    memset(baseAddress, 0, privateKeyData.count)
                }
            }
        }

        guard let privateKey = PrivateKey(data: privateKeyData) else {
            throw EthSignerError.invalidPrivateKey
        }

        // EIP-191 personal sign: "\x19Ethereum Signed Message:\n" + len(message) + message
        let prefix = "\u{19}Ethereum Signed Message:\n\(message.count)"
        var prefixedMessage = Data(prefix.utf8)
        prefixedMessage.append(message)

        // Hash with keccak256
        let hash = Hash.keccak256(data: prefixedMessage)

        // Sign with secp256k1
        guard let signature = privateKey.sign(digest: hash, curve: .secp256k1) else {
            throw EthSignerError.signingFailed
        }

        // Convert signature to KotlinByteArray
        return signature.toKotlinByteArray()
    }

    /// Sign EIP-712 typed data
    /// Matches Android's EthSigner.signTypedData()
    ///
    /// - Parameter typedData: EIP-712 TypedData structure
    /// - Returns: Signature as hex string
    override func __signTypedData(typedData: TypedData) async throws -> String {
        // Get stored private key
        guard let privateKeyData = try? keyManager.getKey() else {
            throw EthSignerError.noStoredKey
        }

        defer {
            // Clear private key from memory
            var mutableKey = privateKeyData
            mutableKey.withUnsafeMutableBytes { ptr in
                if let baseAddress = ptr.baseAddress {
                    memset(baseAddress, 0, privateKeyData.count)
                }
            }
        }

        guard let privateKey = PrivateKey(data: privateKeyData) else {
            throw EthSignerError.invalidPrivateKey
        }

        // Hash according to EIP-712 spec using shared utils
        let hashBytes = Eip712Utils.shared.hashTypedData(hasher: keccak256, typedData: typedData)
        let hash = hashBytes.toNSData()

        // Sign the raw hash (no EIP-191 prefix for EIP-712)
        guard var signature = privateKey.sign(digest: hash, curve: .secp256k1) else {
            throw EthSignerError.signingFailed
        }

        // WalletCore returns v as 0 or 1, but Ethereum expects 27 or 28
        // Adjust the last byte (v value) to match Ethereum format
        if signature.count == 65 {
            let vIndex = signature.count - 1
            let v = signature[vIndex]
            signature[vIndex] = v + 27
        }

        // Convert to hex string
        return signature.toHexString()
    }

    // MARK: - Static Utility Methods (Companion)

    /// Convert mnemonic phrase to private key bytes using BIP39/BIP44
    /// Matches Android's companion object extension: String.mnemonicToKeypair()
    ///
    /// - Parameters:
    ///   - mnemonic: 12 or 24 word mnemonic phrase
    ///   - derivationPath: BIP44 derivation path
    /// - Returns: 32-byte private key
    static func mnemonicToPrivateKey(_ mnemonic: String, derivationPath: String) throws -> Data {
        // Clean up mnemonic
        let cleanMnemonic = mnemonic
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }
            .joined(separator: " ")

        // Validate mnemonic using WalletCore
        guard Mnemonic.isValid(mnemonic: cleanMnemonic) else {
            throw EthSignerError.invalidMnemonic
        }

        // Create HDWallet from mnemonic (BIP39)
        // Note: Using empty passphrase like Android implementation
        guard let wallet = HDWallet(mnemonic: cleanMnemonic, passphrase: "") else {
            throw EthSignerError.invalidMnemonic
        }

        // Derive key using BIP32/BIP44
        // WalletCore uses DerivationPath for path parsing
        guard let derivation = DerivationPath(derivationPath) else {
            throw EthSignerError.keyDerivationFailed
        }

        let privateKey = wallet.getKey(coin: .ethereum, derivationPath: derivation.description)
        return privateKey.data
    }

    /// Derive Ethereum address from private key
    /// Matches Android's companion object extension: ByteArray.privateToAddress()
    ///
    /// - Parameter privateKey: 32-byte private key
    /// - Returns: Ethereum address with 0x prefix
    static func privateKeyToAddress(_ privateKey: Data) throws -> String {
        guard let privKey = PrivateKey(data: privateKey) else {
            throw EthSignerError.invalidPrivateKey
        }

        let publicKey = privKey.getPublicKeySecp256k1(compressed: false)
        let address = AnyAddress(publicKey: publicKey, coin: .ethereum)
        return address.description
    }
}
